src/main/java/org/wildfly/security/pem/Pem.java

```diff
@@ -165,28 +139,76 @@ public static void generatePemContent(ByteStringBuilder target, String type, Byt
         target.append("\n-----END ").append(type).append("-----\n");
     }
 
-    private static X509Certificate parsePemX509CertificateContent(String type, ByteIterator content) throws IllegalArgumentException {
-        if (! type.equals("CERTIFICATE")) {
-            throw log.invalidPemType("CERTIFICATE", type);
+    /**
+     * Returns a {@link PemObject} instance based on a PEM content.
+     *
+     * @param pemContent a {@link CodePointIterator} with the PEM content
+     * @return a {@link PemObject} created from the given PEM content
+     */
+    public static PemObject toPemObject(CodePointIterator pemContent) {
+        pemContent = pemContent.skipCrLf();
+        if (! pemContent.limitedTo(11).contentEquals("-----BEGIN ")) {
+            throw log.malformedPemContent(pemContent.offset());
+        }
+        String type = pemContent.delimitedBy('-').drainToString().trim();
+        final Matcher matcher = VALID_LABEL.matcher(type);
+        if (matcher.find()) {
+            // BEGIN string is 11 chars long
+            throw log.malformedPemContent(matcher.start() + 11);
+        }
+        if (! pemContent.limitedTo(5).contentEquals("-----")) {
+            throw log.malformedPemContent(pemContent.offset());
+        }
+        final byte[] der = pemContent.delimitedBy('-').base64Decode().drain();
```

Comment:
I don't like that we're making copies of bytes over and over again in these routines.  The point of streaming over the byte iterator was that we could iterate any number of PEM objects with a minimum of allocation.  Is there any way we can avoid this?